src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java

```diff
@@ -610,25 +636,37 @@ private boolean jobSetup(Pageable doc) {
 
     // The following methods are CPrinterJob specific.
 
-    @Override
-    @SuppressWarnings("removal")
-    protected void finalize() {
-        synchronized (fNSPrintInfoLock) {
-            if (fNSPrintInfo != -1) {
-                dispose(fNSPrintInfo);
+    static class NSPrintInfoDisposer implements DisposerRecord {
+
+        private long fNSPrintInfo;
+        private Object lock;
+
+        NSPrintInfoDisposer(Object l, long ptr) {
+            lock = l;
+            fNSPrintInfo = ptr;
+        }
+
+
+        public void dispose() {
+            synchronized (lock) {
```

Comment:
I do not think this synchronization is necessary. Before the patch it was used to guard access to fNSPrintInfo field, which might have been lazily initialized and disposed concurrently.

Now we use the Disposer only if fNSPrintInfo has been used and initialized -> we saved it to NSPrintInfoDisposer and from that point it is never changed.

So we can drop the lock, and all check for the field(seems createNSPrintInfo cannot return -1) + mark it as final

Reply:
ok